### PR TITLE
`ApplicationLog` - New Storage implementation

### PR DESCRIPTION
**Pull request by cschuchardt88**
See PR #807 

I close it by accident. Because I had it on my `master` branch 🌿 

---

With this new implementation of "ApplicationLog" you will be easier to expand the "ApplicationLog" for use in neo-node-cli and RPC. Tested with my xunit tests locally. 

_Note that you need to resync the whole blockchain for this new version of "ApplicationLog"_

**Change Log**
- New Storage implementation
- Added new console commands to neo-cli
- Added lookups by block, transaction and contract in neo-cli
- Added Sort methods for by event name

**Storage Size comparison to blockchain**
| Sizes | BlockChain | Application |
| ----   | ------------ | ------------- |
| 2.30GB      | Testnet T5 | Blockchain |
| 2.41GB      | Testnet T5 | ApplicationLog |

### Now you can use ApplicationLog in neo-cli here are the commands

image

### **Output of Command Console for ApplicationLog**

image

image

image

### CODE REVIEW

File: src/ApplicationLogs/LogReader.cs [hunk 1]
```diff
@@ -8,38 +8,55 @@
...
+
+        #region Override Methods
+
         public override void Dispose()
         {
             Blockchain.Committing -= OnCommitting;
             Blockchain.Committed -= OnCommitted;
+            if (Settings.Default.Debug)
+                ApplicationEngine.Log -= OnApplicationEngineLog;
+            GC.SuppressFinalize(this);
```
Comment (vncoelho) [inline src/ApplicationLogs/LogReader.cs:59]: In this safe, everything was disposed? No memory leak?

Comment (cschuchardt88) [reply src/ApplicationLogs/LogReader.cs:59]: No memory leaks. But i do see something that needs changing.

Comment (cschuchardt88) [reply src/ApplicationLogs/LogReader.cs:59]: nevermind, its all good.

**Review by vncoelho — COMMENTED** [review]
image

It looks all good.

**Review by superboyiii — APPROVED** [review]

**Comment by superboyiii** [comment]
@shargon here

**Comment by shargon** [comment]
I need more time to review it

**Comment by roman-khimov** [comment]
It's hard to review, because there are a lot of things mixed here. Storage scheme refactoring and extensions better be separated. I'm mostly worried about API (JSON-RPC) backwards compatibility here.

Do I understand correctly that in the current form there are no new APIs added, only `getapplicationlog` implementation has changed?

**Comment by cschuchardt88** [comment]
@roman-khimov That is correct. Only the way things are stored, with some in additions to the command-line `CLI` commands. The `RPC` functions you see are the same output just different logic.

**Comment by shargon** [comment]
@cschuchardt88 Why fail the format if I saved them with LF?

File: src/ApplicationLogs/Store/LogStorageStore.cs [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+{
+    public sealed class LogStorageStore : IDisposable
+    {
+        #region Prefixes
+
+        private static readonly int Prefix_Size = sizeof(int) + sizeof(byte);
+        private static readonly int Prefix_Block_Trigger_Size = Prefix_Size + UInt256.Length;
+        private static readonly int Prefix_Execution_Block_Trigger_Size = Prefix_Size + UInt256.Length;
+
+        private static readonly int Prefix_Id = 0x414c4f47;                 // Magic Code: (ALOG);
```
Comment (shargon) [inline src/ApplicationLogs/Store/LogStorageStore.cs:31]: What's means ALOG?

File: src/ApplicationLogs/LogReader.cs [hunk 3]
```diff
@@ -50,173 +67,391 @@ protected override void Configure()
...
         [RpcMethod]
         public JToken GetApplicationLog(JArray _params)
         {
-            UInt256 hash = UInt256.Parse(_params[0].AsString());
-            byte[] value = _db.TryGet(hash.ToArray());
-            if (value is null)
-                throw new RpcException(-100, "Unknown transaction/blockhash");
+            if (_params == null || _params.Count == 0)
+                throw new RpcException(-32602, "Invalid params");
+            if (UInt256.TryParse(_params[0].AsString(), out var hash))
```
Comment (shargon) [inline src/ApplicationLogs/LogReader.cs:91]: Allow to call it with index?

Comment (roman-khimov) [reply src/ApplicationLogs/LogReader.cs:91]: That'd be an extension (yes, I will complain about _every_ extension, especially if it's mixed with other code and doesn't have a specific PR for it, it's painfully hard to keep API compatibility when such changes occur randomly, so I'm very happy to see https://github.com/neo-project/neo-modules/pull/865#issuecomment-1885550772 for the current state of PR) and a bit unobvious one, blocks can be indexed, so we usually accept indexes/hashes where blocks are used, but we accept transaction hashes here as well.

**Review by shargon — COMMENTED** [review]
Why it's reverted?

File: src/ApplicationLogs/Store/LogStorageStore.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/ApplicationLogs/Store/LogStorageStore.cs:32]: LOL, 😂  guess? `A` `LOG`! like a log file.

**Comment by cschuchardt88** [comment]
What do you mean @shargon ? I updated file headers and formating https://github.com/neo-project/neo-modules/pull/865/commits/ffa303cc2fdc175073c9a6d185b2e9652be1ad20

**Comment by cschuchardt88** [comment]
nvm i see let me change that

**Comment by cschuchardt88** [comment]
@roman-khimov @shargon @superboyiii @vncoelho 
I forgot to mention #841 is add. But only if `DebugMode` is enabled in the `config.json`. However we did talk about it in the PR before it closed. Issue will have links.

**Comment by shargon** [comment]
@cschuchardt88 I see that we create some Guid for storage enteies, that means that two nodes with the same data won't be able to compare the stored data (as raw), and that's a problem

**Comment by cschuchardt88** [comment]
Ok, it can be changed. But can you explain more in details on what you are or need to do or the reason and what you are trying to compare.

**Comment by cschuchardt88** [comment]
@shargon 
After thinking about it. I have indexing for fast lookups. So do you handle that with other `plugins`? I know we want to use `BlockHash`, `TxHash` or `ContractHash` but a search would be slow.... and DDOS-able

**Comment by shargon** [comment]
> Ok, it can be changed. But can you explain more in details on what you are or need to do or the reason and what you are trying to compare.

@superboyiii ussually did to ensure that the storage didn't change with the changes.

**Comment by cschuchardt88** [comment]
@shargon but this plugin data doesn't really matter? I thought blockchain data like `Transactions`, `Blocks` and `Contract` are only important. Plus with the indexing you can't compare and that limits us on what we can do. This is one reason why lookups are so slow and ddos-able.

**Comment by shargon** [comment]
> @shargon but this plugin data doesn't really matter? I thought blockchain data like `Transactions`, `Blocks` and `Contract` are only important. Plus with the indexing you can't compare and that limits us on what we can do. This is one reason why lookups are so slow and ddos-able.

@superboyiii you only check certains prefixes?

**Comment by cschuchardt88** [comment]
@shargon 

I think @superboyiii only checks `RPC-JSON` data. He already tested and approved, i thought in other PR before closed.

You know how big the database would be if i copy the data to each of these "tables"? Each one is an lookup index for the most part, each bring a different way to look up something. Because if you remember I had a `RPC-JSON` method for `GetContractLog` and its still there but only functions for looking up but no `JSON-RPC`, hints the `CLI` commands.
```csharp
private static readonly int Prefix_Id = 0x414c4f47;                 // Magic Code: (ALOG);
private static readonly byte Prefix_Engine = 0x18;                  // Engine_GUID -> ScriptHash, Message
private static readonly byte Prefix_Engine_Transaction = 0x19;      // TxHash -> Engine_GUID_List
private static readonly byte Prefix_Block = 0x20;                   // BlockHash, Trigger -> NotifyLog_GUID_List
private static readonly byte Prefix_Notify = 0x21;                  // NotifyLog_GUID -> ScriptHash, EventName, StackItem_GUID_List
private static readonly byte Prefix_Contract = 0x22;                // ScriptHash, TimeStamp, EventIterIndex -> txHash, Trigger, NotifyLog_GUID
private static readonly byte Prefix_Execution = 0x23;               // Execution_GUID -> Data, StackItem_GUID_List
private static readonly byte Prefix_Execution_Block = 0x24;         // BlockHash, Trigger -> Execution_GUID
private static readonly byte Prefix_Execution_Transaction = 0x25;   // TxHash -> Execution_GUID
private static readonly byte Prefix_Transaction = 0x26;             // TxHash -> NotifyLog_GUID_List
private static readonly byte Prefix_StackItem = 0xed;               // StackItem_GUID -> Data
```

**Comment by cschuchardt88** [comment]
@shargon @Jim8y This has been tested already by @superboyiii. I deleted branch by mistake cause it was on my master. So i reopened it here.

**Comment by shargon** [comment]
I will review it likely the next week.

**Comment by superboyiii** [comment]
It's good. Let's move on. @shargon

**Review by vncoelho — APPROVED** [review]
I also tested and have been running on neocompiler for some time now.

**Comment by superboyiii** [comment]
@shargon Could you have a review on this?

File: src/ApplicationLogs/ApplicationLogs.csproj [hunk 4]
```diff
@@ -1,9 +1,13 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <PackageId>Neo.Plugins.ApplicationLogs</PackageId>
+    <RootNamespace>Neo.Plugins</RootNamespace>
+    <ImplicitUsings>enable</ImplicitUsings>
   </PropertyGroup>
+  
   <ItemGroup>
-    <ProjectReference Include="..\RpcServer\RpcServer.csproj">
+    <PackageReference Include="Neo.ConsoleService" Version="1.2.0" />
```
Comment (shargon) [inline src/ApplicationLogs/ApplicationLogs.csproj:9]: Use the submodule?

File: src/ApplicationLogs/Store/States/TransactionLogState.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        #region ISerializable
+
+        public virtual int Size =>
+            sizeof(uint) +
+            NotifyLogIds.Sum(s => s.ToByteArray().GetVarSize());
+
+        public virtual void Deserialize(ref MemoryReader reader)
+        {
+            // It should be safe because it filled from a transaction's notifications.
+            uint aLen = reader.ReadUInt32();
```
Comment (shargon) [inline src/ApplicationLogs/Store/States/TransactionLogState.cs:36]: `ReadVarInt` and `WriteVarInt` in all of them?

Comment (cschuchardt88) [reply src/ApplicationLogs/Store/States/TransactionLogState.cs:36]: I do it this way so; there is no extra byte for type of `Integer`.

**Review by vncoelho — COMMENTED** [review]
I have been running this branch for sometime without problems.
However, we need @superboyiii to re-ensure that there is not state problems and I think it is ok.

**Review by shargon — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

**Comment by cschuchardt88** [comment]
😄 My 1st PR merged!

**Pull request merged**